Auth storage

ADmad opened the pull request:
Refs #6628 #6655

`AuthComponent::$_user` and `AuthComponent::sessionKey` are gone. Is it necessary to add some shim for `AuthComponent::sessionKey` for BC or just document the fact that you now need to configure the storage instead.

ADmad committed:
Add Auth StorageInterface and classes

ADmad committed:
Update AuthComponent to use storage class

ADmad committed:
Use MemoryStorage instead of AuthComponent::

ADmad committed:
Remove unneeded NullStorage

ADmad committed:
Update docblocks

ADmad committed:
Add test for SessionStorage

AD7six commented on src/Auth/Storage/SessionStorage.php:
```diff
@@ -0,0 +1,96 @@
+     */
+    public function get()
+    {
+        if (!$this->_session->check($this->_config['key'])) {
```
I think this class should also use a local variable for the user data, and not read from the session on every call

ADmad replied:
I agree, will update.

dereuromark commented:
Is that some preparation for https://github.com/cakephp/cakephp/issues/6247 as well? So that it will be easier to move the auth access towards the request handler?

ADmad committed:
Avoid hitting session each time

markstory commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -170,21 +172,11 @@ class AuthComponent extends Component
      *
-     * @var string
-     */
-    public $sessionKey = 'Auth.User';
```
We need to figure out a BC way to handle this I think.

markstory replied:
We could use `__get()` :smile:

ADmad replied:
We could, though getting isn't the real issue, what to do if user sets it is :smile:

Should the SessionStorage config be updated if $sessionKey is set to new value?

markstory replied:
If we want everything to be totally BC then yeah it probably should. As an end developer I would expect that the code I had in 3.0 would continue to work the same way in 3.1.

markstory commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -797,6 +783,34 @@ public function constructAuthenticate()
         return $this->_authenticateObjects;
     }
 
+    public function storage(StorageInterface $storage = null)
```
Docblocks please.

ADmad replied:
Missed one :stuck_out_tongue:

markstory commented on tests/TestCase/Auth/Storage/SessionStorageTest.php:
```diff
@@ -0,0 +1,92 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         2.0.0
```
Liles.

ADmad commented:
@dereuromark I will help towards that.

ADmad committed:
Add docblock

ADmad committed:
Fix record removal from session storage

ADmad committed:
Add magic methods for BC for removed AuthComponent::$sessionKey

ADmad committed:
Rename _storageObject to _storage.

ADmad commented:
@markstory I add magic methods for $sessionKey BC.

Thinking of renaming `StorageInterface` method to read(), write(), delete() to be consistent with names used by Cache and Session classes.

markstory commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -798,6 +784,81 @@ public function constructAuthenticate()
+
+            $this->config('storage', 'Session');
+            $this->storage()->config('key', $value);
+            return;
```
Nice :+1:

markstory commented on tests/TestCase/Controller/Component/AuthComponentTest.php:
```diff
@@ -1378,4 +1378,21 @@ public function testStatelessFollowedByStatefulAuth()
+        $this->assertEquals('Auth.User', $this->Auth->sessionKey);
+
+        $this->Auth->sessionKey = 'Auth.Member';
+        $this->assertEquals('Auth.Member', $this->Auth->sessionKey);
```
Should you check the value of `storage->config('key')` as well?

ADmad committed:
Add assert

AD7six commented on src/Auth/Storage/SessionStorage.php:
```diff
@@ -0,0 +1,108 @@
+     */
+    public function get()
+    {
+        if ($this->_user) {
```
Would it would be a good idea to account for multiple calls when there is no auth user?

```
public function get()
{
    if ($this->_user !== null) {
        return $this->_user;
    }

    $this->_user = $this->_session->read($this->_config['key']) ?: false;
    return $this->_user;
}
```

That way multiple calls, if there is no user, won't touch the session each time.

ADmad replied:
I did think about doing the same but often people directly manipulate the session value, in which case the above code wont reflect the change.

AD7six replied:
It'd be consistent behavior though. Otherwise doing that (arguably users simply shouldn't do that) sometimes works and sometimes doesn't depending on circumstance and execution order.

ADmad replied:
Right, once `$this->_user` does get populated with record from session directly changing session value won't be reflected anyway.

I'll update.

ADmad committed:
Improve code to avoid accessing session

ADmad committed:
Fix CS errors

ADmad committed:
Set to null instead of unsetting

ADmad committed:
Rename SessionStorageInterface methods.

Makes them consistent with names used by Cache and Session classes.

ADmad commented:
This is ready to go.

AD7six commented:
:+1:

lorenzo commented on src/Controller/Component/AuthComponent.php:
```diff
@@ -630,32 +620,25 @@ public function logout()
         $user = (array)$this->user();
         $this->dispatchEvent('Auth.logout', [$user]);
-        $this->session->delete($this->sessionKey);
         $this->session->delete('Auth.redirect');
```
I think the goal of this features is to be able to avoid the the session entirely, so this line should probably be moved to the other storage as well.

ADmad replied:
So add a `redirectUrl()` method to `SessionStorageInterface`?

markstory replied:
That works for me.

ADmad replied:
Okay, not too thrilled about this but I don't have a better idea.

lorenzo commented:
I think this would require adding a method to `IntegrationTestCase` to simulate having a user logged in in the right storage.

ADmad commented:
> I think this would require adding a method to IntegrationTestCase to simulate having a user logged in in the right storage.

Does it?  For session storage won't just populating the session be enough? For other stateless authenticators you would populate relevant request params and configure Auth to use MemoryStorage.

markstory commented on src/Auth/Storage/SessionStorage.php:
```diff
@@ -0,0 +1,111 @@
+     * @param \Cake\Network\Request $request Request instance.
+     * @param array $config Configuration list.
+     */
+    public function __construct(Request $request, array $config = [])
```
Should these classes also get the response? I could see a world where someone would want to do an encrypted cookie 'session' for the user data.

ADmad committed:
Add StorageInterface::redirectUrl().

This move redirect url reading/writing to auth storage classes
and complete decoupling of AuthComponent with session.

ADmad committed:
Add Response as argument for auth storage classes.

ADmad committed:
Make redirect url key configurable.

ADmad committed:
Change "url" to "URL".

Shout out to @dereuromark :)

ADmad commented:
- Added `StorageInterface::redirectUrl()`.
- Passed `Response` to constructor of storage classes.

markstory commented on tests/TestCase/Auth/Storage/SessionStorageTest.php:
```diff
@@ -0,0 +1,105 @@
+        $this->response = new Response();
+        $this->storage = new SessionStorage($this->request, $this->response, ['key' => 'Auth.AuthUser']);
+        $this->user = ['id' => 1];
+    }
```
redirectUrl() has no tests.

ADmad replied:
Added.

ADmad committed:
Add tests for SessionStorage::redirectUrl()

dereuromark commented:
Looks great.

markstory commented:
Looks good. @ADmad do you need a hand writing the docs?

ADmad commented:
@markstory I have opened PR cakephp/docs#2817. Please improve upon it as you see fit.
